src/interface/variable_pack.hpp

```diff
@@ -0,0 +1,234 @@
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//=======================================================================================
+#ifndef INTERFACE_VARIABLE_PACK_HPP_
+#define INTERFACE_VARIABLE_PACK_HPP_
+
+#include <array>
+#include <forward_list>
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "interface/metadata.hpp"
+#include "interface/variable.hpp"
+
+namespace parthenon {
+
+// some convenience aliases
+namespace vpack_types {
+template <typename T>
+using VarList = std::forward_list<std::shared_ptr<CellVariable<T>>>;
+using IndexPair = std::pair<int, int>;
+using StringPair = std::pair<std::vector<std::string>, std::vector<std::string>>;
+} // namespace vpack_types
+
+using PackIndexMap = std::map<std::string, vpack_types::IndexPair>;
+template <typename T>
+using ViewOfParArrays = ParArrayND<ParArray3D<T>>;
+
+// Try to keep these Variable*Pack classes as lightweight as possible.
+// They go to the device.
+template <typename T>
+class VariablePack {
+ public:
+  VariablePack() = default;
+  VariablePack(const ViewOfParArrays<T> view, const std::array<int, 4> dims)
+      : v_(view), dims_(dims) {}
+  KOKKOS_FORCEINLINE_FUNCTION
+  ParArray3D<T> &operator()(const int n) const { return v_(n); }
+  KOKKOS_FORCEINLINE_FUNCTION
+  T &operator()(const int n, const int k, const int j, const int i) const {
+    return v_(n)(k, j, i);
+  }
+  KOKKOS_FORCEINLINE_FUNCTION
+  int GetDim(const int i) {
+    assert(i > 0 && i < 5);
+    return dims_[i - 1];
+  }
+
+ protected:
+  ViewOfParArrays<T> v_;
+  std::array<int, 4> dims_;
+};
+
+template <typename T>
+class VariableFluxPack : public VariablePack<T> {
+ public:
+  VariableFluxPack() = default;
+  VariableFluxPack(const ViewOfParArrays<T> view, const ViewOfParArrays<T> f0,
+                   const ViewOfParArrays<T> f1, const ViewOfParArrays<T> f2,
+                   const std::array<int, 4> dims, const int nflux)
+      : VariablePack<T>(view, dims), f_({f0, f1, f2}), nflux_(nflux) {}
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  ViewOfParArrays<T> &flux(const int dir) const { return f_[dir]; }
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  T &flux(const int dir, const int n, const int k, const int j, const int i) const {
+    return f_[dir](n)(k, j, i);
+  }
+
+ private:
+  std::array<ViewOfParArrays<T>, 3> f_;
+  int nflux_;
+};
+
+// mapping to a tuple instead of using multiple maps reduces # of lookups
+// this wouldn't be super important if lookup time was constant,
+// but std::maps are trees, not hash tables and have an O(log(N)) lookup.
+// Unfortunately, std::pair doesn't work. So I have to roll my own.
+// I have no idea why std::pair doesn't work.
+// It appears to be an interaction between caused by a std::map<key,std::pair>
+// Possibly it's a compiler bug. gcc/7.4.0
+// ~JMM
```

Comment:
What about `std::unordered_map`? That has O(1) lookup on average (worst case O(N)).

Reply:
Also would 'map' well to Kokkos::unordered_map, although Kokkos::unordered_map pushes the onus of finding the entry on the user.

Reply:
This is a good idea, so I gave it a shot. Unfortunately, the key, `std::vector<std::string>` is not a hashable type but is a comparable type. So `map` works but `unordered_map` would require defining a custom hashing function. I'm going to stick with `map` for now.

We can revisit if it's an issue.